File: csun_catalog_and_schedules_bot/bot.py
# bot.py
import os
import urllib3
import json
import discord
from dotenv import load_dotenv
import time
import mysql.connector
from mysql.connector import errorcode


load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')

client = discord.Client()


def show_classes(subject, number):
    sems = ["Fall", "Spring"]
    json_blobs = []
    
    for sem in sems:    
        url = u"https://api.metalab.csun.edu/curriculum/api/2.0/terms/" + sem + "-2022/courses/" + subject
        #print("\n Data Link: " + url)

        # try to read the data and load
        try:
            data = json.loads(urllib3.PoolManager().request("GET", url).data)
        except Exception as e:
            data = json.loads({})

        for course in data["courses"]:
            if (number == course["catalog_number"]):
                json_blobs.append(course)
                break
            
    if len(json_blobs) > 0:
        ret = str(json_blobs[0]["subject"].upper() + " " + json_blobs[0]["catalog_number"] + " " + json_blobs[0]["title"] + "\n\n")
        for a in json_blobs:
            if a["description"] is not None:
                ret += a["description"]
                break
        return ret




def show_schedule(sem, year, sub, code):
    if sem.lower() == "spring" and year == "2023":
        
        #data = json.load(open("storedschedules/" + sub.upper() + "_schedule.json"))
        load_dotenv()
        dausername = os.getenv('sqlusername')
        dapassword = os.getenv('sqlpass')
        request_tuple = (sub, code)
        data = {}
        try:
            rootConnection = mysql.connector.connect(
            user=dausername,
            password=dapassword,
            host='127.0.0.1',
            database='csun')
    
            rootCursor = rootConnection.cursor()
    
        except mysql.connector.Error as err:
            if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:
                print('Invalid credentials')
            elif err.errno == errorcode.ER_BAD_DB_ERROR:
                print('Database not found')
            else:
                print('Cannot connect to database:', err)

        else:
            rootCursor.execute('select * from section where subject_code = %s and catalog_number = %s', request_tuple)
            stuffs = rootCursor.fetchall()
            
            for row in stuffs:
                temp_dict = {}
                temp_dict["catalog_number"] = row[1]
                temp_dict["title"] = row[9]
                temp_dict["meetings"] = []
                temp_dict["meetings"].append({})
                temp_dict["meetings"][0]["days"] = row[7]
                temp_dict["meetings"][0]["location"] = row[4]
                temp_dict["meetings"][0]["start_time"] = row[5]
                temp_dict["meetings"][0]["end_time"] = row[6]
                temp_dict["class_number"] = str(row[3])
                temp_dict["enrollment_cap"] = row[8]
                temp_dict["enrollment_count"] = 0
                temp_dict["instructors"] = []
                temp_dict["instructors"].append({})
                temp_dict["instructors"][0]["instructor"] = row[2]
                data["classes"].append(temp_dict)
            rootConnection.close()

                
        
    else:
        url = u"https://api.metalab.csun.edu/curriculum/api/2.0/terms/" + sem + "-" + \
                                                                          year + "/classes/" + \
                                                                          sub
            # try to read the data and load
        try:
            data = json.loads(urllib3.PoolManager().request("GET", url).data)
        except Exception as e:
            data = json.loads({})

    def find_class(current_class):
        ret_value = ""
        for course in data["classes"]:
            if (current_class == course["catalog_number"]):
                ret_value = course["title"]
        return ret_value

    blob_list = []
    curr_time = time.asctime(time.localtime(time.time())).split()
    blob_list.append(sub.upper() + " " + code + " " +
                     find_class(code) + " - " + sem.upper() + " " + year + 
                     " - As of " + curr_time[0] + " " + curr_time[2] + " " + curr_time[1] + " "  + curr_time[4] + " " + curr_time[3])
    blob_list.append(
        "\n\tSection\t\tLocation\tDays\t\t Seats Aval\t\t\t  Time\t\t\t\tFaculty")
    blob_list.append(
        "\t-------\t\t--------\t----\t\t ----------\t\t\t  ----\t\t\t\t-------")

    for course in data["classes"]:
        # if a class has no meetings, it should not be on schedule
        if (len(course["meetings"]) > 0) and code == course["catalog_number"]:
            section_string = []
          
            section_string.append("\t " + course["class_number"] + " ")
            
            if (len(course["meetings"][0]["location"]) == 3):
                
                section_string.append("  ")
            
            # Location 
            if (len(course["meetings"][0]["location"]) != 7):
                # (JD1600A is one character longer than all other class location strings, so it messes up tabs)
                section_string.append("\t\t" + course["meetings"][0]["location"])
                
            else:
                section_string.append("\t   " + course["meetings"][0]["location"])
                
            # Days
            if len(str(course["meetings"][0]["days"])) == 1:
                section_string.append("\t  " + str(course["meetings"][0]["days"]) + "  ")

            elif len(str(course["meetings"][0]["days"])) == 2:
                section_string.append("\t " + str(course["meetings"][0]["days"]) + "  ")
            
            elif len(str(course["meetings"][0]["days"])) == 3:
                section_string.append("\t " + str(course["meetings"][0]["days"]) + " ")
                
            elif str(course["meetings"][0]["days"]) == "None":
                section_string.append("\t --  ")
                
            else:
                section_string.append("\t" + str(course["meetings"][0]["days"]) + " ")
                # print(str(course["meetings"][0]["days"]))

            
            # Seats Available
            if len(str(course["enrollment_cap"] - course["enrollment_count"])) == 1:
                section_string.append("\t\t    " + str(course["enrollment_cap"] - course["enrollment_count"]))
                
            else:
                section_string.append("\t\t   " + str(course["enrollment_cap"] - course["enrollment_count"]))

            # Time 
            section_string.append("\t\t    " +
                                  (course["meetings"][0]["start_time"])[0:2] + ":" +
                                  (course["meetings"][0]["start_time"])[2:4]
                                  + " - " +
                                  (course["meetings"][0]["end_time"])[0:2] + ":" +
                                  (course["meetings"][0]["end_time"])[2:4])

            # Instructor
            # if a class has no instructor, print Staff instead
            if (len(course["instructors"]) > 0) and course["instructors"][0]["instructor"] != "Staff":
                section_string.append("\t" + course["instructors"][0]["instructor"])
            else:
                section_string.append("\t\t\t" + "Staff")

            blob_list.append(" ".join(section_string))
            print("------------------------------------------------------------------")
            print(section_string)
    return "\n".join([str(x) for x in blob_list])


@client.event
async def on_ready():
    print(f'{client.user} has connected to Discord!')


@client.event
async def on_message(message):
    if message.author == client.user:
        return

    msg_split = message.content.split()
    print(message.author, end="")
    print(" [" + message.content + "]")

    if message.content.__contains__("!csun") and len(msg_split) == 3:
        response1 = show_classes(msg_split[1], msg_split[2])
        response2 = show_schedule("Fall", "2022", msg_split[1], msg_split[2])
        await message.channel.send("```" + str(response1) + "\n\n" + str(response2) + "```")
        
    elif message.content.__contains__("!csun grade"):
        grade = 0
        total_weight = 0
        for i in range(2, int(len(msg_split)-2/2)+1, 2):
            grade += (float(msg_split[i]) * (float(msg_split[i+1]))/100)
            total_weight += float(msg_split[i+1])
            #await message.channel.send("```" + str(i) + "```")
        await message.channel.send("```" + str(grade) + "```")
        if total_weight != 100.0:
            await message.channel.send("```Warning: Weights don't total 100.```")
            
    elif len(msg_split) > 3 and message.content.__contains__("!csun"):
        response1 = show_classes(msg_split[1], msg_split[2])
        response2 = show_schedule(msg_split[3], "20" + msg_split[4], msg_split[1], msg_split[2])
        await message.channel.send("```" + str(response1) + "\n\n" + str(response2) + "```")

    elif message.content.__contains__("!csun help"):
        await message.channel.send("```Shows both class description and schedule by default. Default schedule is Fall 2022 \
                                   \nTo show different schedule, append it to the end.\n\n" +
                                   "For default:\n\t!csun subject class_code\nExample:\n\t!csun comp 182\n\n" + 
                                   "For Different Semester:\n\t!csun subject class_code semester YY\n" +
                                   "Example:\n\t!csun subject class_code spring 23\n\n" + 
                                   "For Grade:\n\t!csun grade (grade weight)*\n"  + 
                                   "Example:\n\t!csun grade 74 25 85 35 70 40\n\nSource Code: \
                                   \nhttps://github.com/kyeou/Python-Scripts/tree/main/csun_catalog_and_schedules_bot```")
        
    
        

client.run(TOKEN)

File: csun_catalog_and_schedules_bot/bot.js
require('dotenv').config()
const {
  token
} = require('./config.json');
const {
  Client,
  GatewayIntentBits,
  GatewayDispatchEvents,
  EmbedBuilder
} = require('discord.js');
const client = new Client({
  intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages, GatewayIntentBits.MessageContent]
});
client.on('ready', () => {
  console.log(`Logged in as ${client.user.tag}!`);
});

function show_prof(subject, prof, itchid) {
  const class_codes = ["AE", "AM", "AAS", "ACCT", "AFRS", "AIS", "ANTH", "ARAB", "ARMN", "ART", "ASTR", "AT", "ATHL", "BANA", "BIOL", "BLAW", "BUS", "CE", "CADV", "CAS", "CCE", "CD", "CECS", "CHS", "CHEM", "CHIN", "CIT", "CJS", "CLAS", "CM", "COMP", "COMS", "CTVA", "DEAF", "EED", "ECE", "ECON", "EDUC", "ELPS", "ENGL", "ENT", "EOH", "EPC", "FCFC", "FCHC", "FCS", "FIN", "FLIT", "FREN", "GBUS", "GEOG", "GEOL", "GWS", "HEBR", "HHD", "HIST", "HSCI", "HUM", "INDS", "IS", "ITAL", "JS", "JAPN", "JOUR", "KIN", "KNFC", "KOR", "LIB", "LING", "LRS", "ME", "MATH", "MCOM", "MGT", "MKT", "MSE", "MUS", "NURS", "PERS", "PHIL", "PHSC", "PHYS", "POLS", "PSY", "PT", "QS", "RS", "RE", "RTM", "RUSS", "SED", "SCI", "SCM", "SOC", "SOM", "SPAN", "SPED", "SUS", "SUST", "SWRK", "TH", "UNIV", "URBS"];
  if (class_codes.includes(subject.toUpperCase())) {
    var ret2 = "";
    console.log(subject + " " + prof);
    console.log("Show prof called.");
    var ret1 = "";
    require("request")({
      url: `http://127.0.0.1:8000/${subject}/prof`,
      json: true
    }, async function (error, response, body) {
      
      console.log(`http://127.0.0.1:8000/${subject}/prof`);
      if (!error) {
        const stuffs = JSON.parse(JSON.stringify(body));
        var prof_email = "";
        stuffs.profs.forEach(da_prof => {
          if (da_prof.includes(prof.toLowerCase()) && !prof_email) {
            prof_email = da_prof;
          }
        });
        if (prof_email) {
          console.log(prof_email);
          ret2 = ret2.concat("Subject: " + subject.toUpperCase() + " - Fall 2022\nProfessor: " + prof_email + "\n");
          ret2 = ret2.concat("\n\tSection\t Class\t\t Location\t\tDays\t\t  Seats\t\t\t  Time\n\t-------\t-------\t\t--------\t\t----\t\t  -----\t\t\t  ----\n");
          stuffs[prof_email].classes.forEach(course => {
            ret2 = ret2.concat("\t " + course.class_number);
            ret2 = (course.catalog_number.length === 4) ? ret2.concat("\t   " + course.catalog_number) : ret2.concat("\t   " + course.catalog_number + " ");
            ret2 = (course.meetings[0].location.length === 3) ? ret2.concat("   ") : ret2.concat("");
            ret2 = (course.meetings[0].location.length === 5) ? ret2.concat("\t\t   " + course.meetings[0].location) : ret2.concat("\t\t  " + course.meetings[0].location);

            switch (course.meetings[0].days.length) {
              case 1:
                ret2 = ret2.concat("\t\t   " + course.meetings[0].days);
                break;
              case 2:
                ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
                break;
              case 3:
                ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
                break;
              default:
                ret2 = ret2.concat("\t\t " + course.meetings[0].days);
            }


            ret2 = ret2.concat(`\t\t\t ${(course.enrollment_cap - course.enrollment_count)}\t\t\t${course.meetings[0].start_time.substring(0, 2)}:${course.meetings[0].start_time.substring(2, 4)} - ${course.meetings[0].end_time.substring(0, 2)}:${course.meetings[0].end_time.substring(2, 4)}`);
            ret2 = ret2.concat("\n");
          });
        } else {
          ret2 = ret2.concat(`${prof.substring(0, 1).toUpperCase()}${prof.substring(1).toLowerCase()} not teaching ${subject.toUpperCase()} in Fall 2022.`);
        }
      }
    }); /*end request*/
    if (ret2.length < 4000) {
      setTimeout(async () => {
        await client.channels.cache.get(itchid).send("```" + ret2 + "```")
      }, 2000);
    } else {
      setTimeout(async () => {
        await client.channels.cache.get(itchid).send("```Response too long.```")
      }, 2000);
    }
  } else {
    ret_error = `${subject} is not a subject.`;
    setTimeout(async () => {
      await client.channels.cache.get(itchid).send("```Not a subject.```")
    }, 2000);
  }
}

function show_levels(subject, level, itchid) {
  console.log("Show levels called.");
  var ret1 = "";
  require("request")({
    url: `http://127.0.0.1:8000/${subject}/catalog`,
    json: true
  }, async function (error, response, body) {
    if (!error) {
      ret1 = ret1.concat(subject.toUpperCase() + " " + level + "-level classes\n")
      for (let i = 0; i < subject.length; i++) {
        ret1 = ret1.concat("-")
      }
      ret1 = ret1.concat("\n")
      console.log(`http://127.0.0.1:8000/${subject}/catalog`);
      const stuffs = JSON.parse(JSON.stringify(body));
      stuffs.forEach(course => {
        if (course.catalog_number[0] === level[0]) {
          ret1 = ret1.concat(`${course.catalog_number} - ${course.title}\n`);
        }
      });
    }
  }); /*end request*/
  setTimeout(async () => {
    await client.channels.cache.get(itchid).send("```" + ret1 + "```")
  }, 2000);
}

function show_class(subject, code, itchid) {
  console.log("Show class called.");
  var ret1 = "",
    ret2 = "";
  require("request")({
    url: `http://127.0.0.1:8000/${subject}/catalog`,
    json: true
  }, async function (error, response, body) {
    console.log(`http://127.0.0.1:8000/${subject}/catalog`);
    if (!error) {
      const stuffs = JSON.parse(JSON.stringify(body));
      stuffs.forEach(course => {
        if (course.catalog_number === code) {
          ret1 = ret1.concat(`${course.subject} ${course.catalog_number} ${course.title}\n\n${course.description}\n\n${course.subject} ${course.catalog_number} ${course.title}`);
          ret1 = ret1.concat(" - FALL 2022");
          require("request")({
            url: `http://127.0.0.1:2222/time`,
            json: true
          }, async function (error, response, body) {
            console.log(`http://127.0.0.1:2222/time`);
            if (!error) {
              ret1 = ret1.concat(body + "\n")
            }
          });
        }
      });
    }
  }); /*end request*/
  require("request")({
    url: 'https://api.metalab.csun.edu/curriculum/api/2.0/terms/Fall-2022/classes/' + subject,
    json: true
  }, async function (error, response, body) {
    console.log('https://api.metalab.csun.edu/curriculum/api/2.0/terms/Fall-2022/classes/' + subject);
    if (!error) {
      const stuffs = JSON.parse(JSON.stringify(body));
      ret2 = ret2.concat("\n\tSection\t\tLocation\t\tDays\t\t  Seats\t\t\t  Time\t\t\t\t\tFaculty\n\t-------\t\t--------\t\t----\t\t  -----\t\t\t  ----\t\t\t\t\t-------\n");
      stuffs.classes.forEach(course => {
        if (course.catalog_number === code && course.meetings.length > 0) {
          ret2 = ret2.concat("\t " + course.class_number);
          ret2 = (course.meetings[0].location.length === 3) ? ret2.concat("\t\t       ") : ret2.concat("");
          ret2 = (course.meetings[0].location.length === 5) ? ret2.concat("\t\t   " + course.meetings[0].location) : ret2.concat("\t\t  " + course.meetings[0].location);



          switch (course.meetings[0].days.length) {
            case 1:
              ret2 = ret2.concat("\t\t   " + course.meetings[0].days);
              break;
            case 2:
              ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
              break;
            case 3:
              ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
              break;
            default:
              ret2 = ret2.concat("\t\t " + course.meetings[0].days);
          }

          ret2 = ret2.concat(`\t\t\t ${(course.enrollment_cap - course.enrollment_count)}\t\t\t${course.meetings[0].start_time.substring(0, 2)}:${course.meetings[0].start_time.substring(2, 4)} - ${course.meetings[0].end_time.substring(0, 2)}:${course.meetings[0].end_time.substring(2, 4)}`);
          ret2 = (course.instructors.length > 0) ? ret2.concat("\t\t" + course.instructors[0].instructor) : ret2.concat("\t\t\t\tStaff");
          ret2 = ret2.concat("\n");
        }
      });
    } /*end if*/
    if (ret1.length + ret2.length < 4000) {
      setTimeout(async () => {
        await client.channels.cache.get(itchid).send("```" + ret1 + ret2 + "```")
      }, 2000);
    } else {
      setTimeout(async () => {
        await client.channels.cache.get(itchid).send("```Response too long.```")
      }, 2000);
    }
  }); /*end request*/
}

function show_class_with_term(subject, code, semester, year, itchid) {
  console.log("Show class override called.");
  if (semester == "spring" && year == '23') {
    var ret1 = "",
      ret2 = "";
    require("request")({
      url: `http://127.0.0.1:8000/${subject}/catalog`,
      json: true
    }, async function (error, response, body) {
      console.log(`http://127.0.0.1:8000/${subject}/catalog`);
      if (!error) {
        const stuffs = JSON.parse(JSON.stringify(body));
        stuffs.forEach(course => {
          if (course.catalog_number === code) {
            ret1 = ret1.concat(course.subject + " " + course.catalog_number + " " + course.title + "\n\n" + course.description + "\n\n" + course.subject + " " + course.catalog_number + " " + course.title);
            ret1 = ret1.concat(" - " + semester.toUpperCase() + " " + year);
            require("request")({
              url: `http://127.0.0.1:2222/time`,
              json: true
            }, async function (error, response, body) {
              console.log(`http://127.0.0.1:2222/time`);
              if (!error) {
                ret1 = ret1.concat(body + "\n")
              }
            })
          }
        });
      }
    }); /*end request*/
    require("request")({
      url: `http://127.0.0.1:8000/${subject}/schedule`,
      json: true
    }, async function (error, response, body) {
      console.log(`http://127.0.0.1:8000/${subject}/schedule`);
      if (!error) {
        const stuffs = JSON.parse(JSON.stringify(body));
        ret2 = ret2.concat("\n\tSection\t\tLocation\t\tDays\t\t  Seats\t\t\t  Time\t\t\t\t\tFaculty\n\t-------\t\t--------\t\t----\t\t  -----\t\t\t  ----\t\t\t\t\t-------\n");
        stuffs.classes.forEach(course => {
          if (course.catalog_number === code && course.meetings.length > 0) {
            ret2 = (course.class_number.length === 5) ? ret2.concat("\t " + course.class_number) : ret2.concat("\t  " + course.class_number);
            ret2 = (course.meetings[0].location.length === 3) ? ret2.concat("   ") : ret2.concat("");
            ret2 = (course.meetings[0].location.length === 5) ? ret2.concat("\t\t   " + course.meetings[0].location) : ret2.concat("\t\t  " + course.meetings[0].location);
            if (course.meetings[0].days !== null) {

              switch (course.meetings[0].days.length) {
                case 1:
                  ret2 = ret2.concat("\t\t   " + course.meetings[0].days);
                  break;
                case 2:
                  ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
                  break;
                case 3:
                  ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
                  break;
                default:
                  ret2 = ret2.concat("\t\t " + course.meetings[0].days);
              }

            } else {
              ret2 = ret2.concat("\t\t  --");
            }

           
            ret2 = ret2.concat(`\t\t\t ${(course.enrollment_cap - course.enrollment_count)}\t\t\t${course.meetings[0].start_time.substring(0, 2)}:${course.meetings[0].start_time.substring(2, 4)} - ${course.meetings[0].end_time.substring(0, 2)}:${course.meetings[0].end_time.substring(2, 4)}`);
            ret2 = (course.instructors.length > 0 && course.instructors !== "Staff") ? ret2.concat("\t\t\t" + course.instructors[0].instructor) : ret2.concat("\t\t\t\tStaff");
            ret2 = ret2.concat("\n");
          }
        });
        if (ret1.length + ret2.length < 4000) {
          setTimeout(async () => {
            await client.channels.cache.get(itchid).send("```" + ret1 + ret2 + "```")
          }, 2000);
        } else {
          setTimeout(async () => {
            await client.channels.cache.get(itchid).send("```Response too long.```")
          }, 2000);
        }
      } /*end if*/
    }); /*end request*/
  } else {
    var ret1 = "",
      ret2 = "";
    require("request")({
      url: `http://127.0.0.1:8000/${subject}/catalog`,
      json: true
    }, async function (error, response, body) {
      console.log(`http://127.0.0.1:8000/${subject}/catalog`);
      if (!error) {
        const stuffs = JSON.parse(JSON.stringify(body));
        stuffs.forEach(course => {
          if (course.catalog_number === code) {
            ret1 = ret1.concat(course.subject + " " + course.catalog_number + " " + course.title + "\n\n" + course.description + "\n\n" + course.subject + " " + course.catalog_number + " " + course.title);
            let currentDate = new Date();
            ret1 = ret1.concat(" - " + semester.toUpperCase() + " " + year);
            require("request")({
              url: `http://127.0.0.1:2222/time`,
              json: true
            }, async function (error, response, body) {
              console.log(`http://127.0.0.1:2222/time`);
              if (!error) {
                ret1 = ret1.concat(body + "\n")
              }
            })
          }
        });
      }
    }); /*end request*/
    require("request")({
      url: 'https://api.metalab.csun.edu/curriculum/api/2.0/terms/' + semester + '-20' + year + '/classes/' + subject,
      json: true
    }, async function (error, response, body) {

      console.log('https://api.metalab.csun.edu/curriculum/api/2.0/terms/' + semester + '-20' + year + '/classes/' + subject);

      if (!error) {
        const stuffs = JSON.parse(JSON.stringify(body));
        ret2 = ret2.concat("\n\tSection\t\tLocation\t\tDays\t\t  Seats\t\t\t  Time\t\t\t\t\tFaculty\n\t-------\t\t--------\t\t----\t\t  -----\t\t\t  ----\t\t\t\t\t-------\n");
        stuffs.classes.forEach(course => {
          if (course.catalog_number === code && course.meetings.length > 0) {

            ret2 = ret2.concat("\t " + course.class_number);
            ret2 = (course.meetings[0].location.length === 3) ? ret2.concat("   ") : ret2.concat("");
            ret2 = (course.meetings[0].location.length === 5) ? ret2.concat("\t\t   " + course.meetings[0].location) : ret2.concat("\t\t  " + course.meetings[0].location);



            switch (course.meetings[0].days.length) {
              case 1:
                ret2 = ret2.concat("\t\t   " + course.meetings[0].days);
                break;
              case 2:
                ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
                break;
              case 3:
                ret2 = ret2.concat("\t\t  " + course.meetings[0].days);
                break;
              default:
                ret2 = ret2.concat("\t\t " + course.meetings[0].days);
            }


            ret2 = ret2.concat(`\t\t\t ${(course.enrollment_cap - course.enrollment_count)}\t\t\t${course.meetings[0].start_time.substring(0, 2)}:${course.meetings[0].start_time.substring(2, 4)} - ${course.meetings[0].end_time.substring(0, 2)}:${course.meetings[0].end_time.substring(2, 4)}`);
            ret2 = (course.instructors.length > 0) ? ret2.concat("\t\t" + course.instructors[0].instructor) : ret2.concat("\t\t\t\tStaff");
            ret2 = ret2.concat("\n");

          }
        });


        if (ret1.length + ret2.length < 4000) {
          setTimeout(async () => {
            await client.channels.cache.get(itchid).send("```" + ret1 + ret2 + "```")
          }, 2000);
        } else {
          setTimeout(async () => {
            await client.channels.cache.get(itchid).send("```Response too long.```")
          }, 2000);
        }



      } /*end if*/
    }); /*end request*/
  }
}

client.on('messageUpdate', (oldMessage, newMessage) => {
  if (newMessage.author.tag == "Mysto#8529") {
    client.channels.cache.get("1028088994203967498").send(`[${newMessage.author.tag}] [${newMessage.content}]`)
  }
});

client.on('messageDelete', (message) => {
 // if (message.author.tag == "Mysto#8529") {
    client.channels.cache.get("1028088994203967498").send(`[${message.author.tag}] [${message.content}]`)
  //}
});

client.on('interactionCreate', async interaction => {
  if (!interaction.isChatInputCommand()) return;

  const {
    commandName
  } = interaction;

  if (commandName === 'class') {

    itchid = interaction.channelId;
    semester = interaction.options.getString('semester');
    year = interaction.options.getString('year');

    if ((semester || year) && !(semester && year)) {

      await interaction.reply("Need both semester and year if other than Fall 2022.")

    } else if (semester && year) {

      var subject = "",
        fir_class = "",
        sec_class = "",
        thi_class = "";

      subject = interaction.options.getString('subject').toLowerCase();
      fir_class = interaction.options.getString('catalog_number');
      sec_class = interaction.options.getString('catalog_number1');
      thi_class = interaction.options.getString('catalog_number2');

      show_class_with_term(subject, fir_class, semester, year, itchid);
      if (sec_class) {
        show_class_with_term(subject, sec_class, semester, year, itchid);
      }
      if (thi_class) {
        show_class_with_term(subject, thi_class, semester, year, itchid);
      }


      await interaction.reply("Gimme a sec");


    } else {


      var subject = "",
        fir_class = "",
        sec_class = "",
        thi_class = "";


      subject = interaction.options.getString('subject').toLowerCase();
      fir_class = interaction.options.getString('catalog_number');
      sec_class = interaction.options.getString('catalog_number1');
      thi_class = interaction.options.getString('catalog_number2');


      show_class(subject, fir_class, itchid);
      if (sec_class) {
        show_class(subject, sec_class, itchid);
      }
      if (thi_class) {
        show_class(subject, thi_class, itchid);
      }

      await interaction.reply("Gimme a sec");
    }
  } else if (commandName === 'classes') {

    itchid = interaction.channelId;

    const class1 = interaction.options.getString('class1').split(" ")
    const class2 = (interaction.options.getString('class2')) ? interaction.options.getString('class2').split(" ") : [];
    const class3 = (interaction.options.getString('class3')) ? interaction.options.getString('class3').split(" ") : [];


    var semester = "",
      year = "";

    semester = interaction.options.getString('semester');
    year = interaction.options.getString('year');


    if ((semester || year) && !(semester && year)) {

      await interaction.reply("Need both semester and year if other than Fall 2022.");

    } else if (semester && year) {

      show_class_with_term(class1[0], class1[1], semester, year, itchid);
      if (class2.length) {
        show_class_with_term(class2[0], class2[1], semester, year, itchid);
      }
      if (class3.length) {
        show_class_with_term(class3[0], class3[1], semester, year, itchid);
      }

      await interaction.reply("Gimme a sec");

    } else {

      show_class(class1[0], class1[1], itchid);
      if (class2.length) {
        show_class(class2[0], class2[1], itchid);
      }
      if (class3.length) {
        show_class(class3[0], class3[1], itchid);
      }

      await interaction.reply("Gimme a sec");
    }
  } else if (commandName === 'prof') {

    itchid = interaction.channelId;
    show_prof(interaction.options.getString('subject'), interaction.options.getString('prof_name'), itchid);
    await interaction.reply("Gimme a sec");

  } else if (commandName === 'help') {

    let ret = "```\"/class\" for 1 or more classes of common subject \n\n" +
      "\"/classes\" for 1 or more classes of different subjects \n\n" +
      "\"/prof\" to show a prof's teaching schedule \n\n" +
      "Source Code:\nhttps://github.com/kyeou/Python-Scripts/tree/main/csun_catalog_and_schedules_bot```";

    await interaction.reply(ret);

  } else if (commandName === 'level') {

    itchid = interaction.channelId;
    show_levels(interaction.options.getString('subject'), interaction.options.getString('level'), itchid)
    await interaction.reply("Gimme a sec");

  }
});
client.login(token);